### PR TITLE
Refactor Generate Function

### LINKED ISSUE #63: Restructure Generate Function
Currently the main generate function is all in one big wall of text. It would be ideal to break this function into smaller functions to improve readability

### PR DESCRIPTION
**Pull request by Sheikah45**
Fixes #63. Changes generate function for readability and utilization of the binary masks whenever possible

### CODE REVIEW

File: src/java/generator/MapGenerator.java [hunk 1]
```diff
@@ -21,33 +21,58 @@
...
 
 	//read from cli args
-	private static String FOLDER_PATH = ".";
-	private static String MAP_NAME = "debugMap";
-	private static long SEED = SEED_GEN.nextLong();
-	private static Random RANDOM = new Random(SEED);
-	private static Optional<Biome> BIOME = Optional.empty();
+	private static String folderPath = ".";
+	private static String mapName = "debugMap";
+	private static long seed = SEED_GEN.nextLong();
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java]: I'd remove the SEED_GEN and just create a new Random() inline.

File: src/java/generator/MapGenerator.java [hunk 2]
```diff
@@ -93,21 +118,80 @@ public void save(String folderName, String mapName, SCMap map) {
...
+		System.out.printf("Using biome %s\n", biomeSet.getName());
+		map.biome.setTerrainMaterials(biomeSet.getTerrainMaterials());
+		map.biome.setWaterSettings(biomeSet.getWaterSettings());
+		map.biome.setLightingSettings(biomeSet.getLightingSettings());
+
+		Preview.generate(
+			map.getPreview(),
+			map,
+			rock.getBinaryMask()
+		);
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java]: remove newlines

File: src/java/generator/MapGenerator.java [hunk 3]
```diff
@@ -333,67 +359,67 @@ private static void interpretArguments(Map<String, String> arguments) {
 			System.exit(0);
 		}
 
-		if (!Arrays.asList("folder-path").stream().allMatch(arguments::containsKey) && !DEBUG) {
+		if (!Stream.of("folder-path").allMatch(arguments::containsKey) && !DEBUG) {
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java]: This is a stream of a single argument. Just call `arguments.containsKey("folder-path")`.

File: src/java/generator/MapGenerator.java [hunk 4]
```diff
@@ -333,67 +359,67 @@ private static void interpretArguments(Map<String, String> arguments) {
...
-			RAMP_DENSITY = (float) StrictMath.round(RAMP_DENSITY*127)/127;
+			rampDensity = StrictMath.min(Float.parseFloat(arguments.get("ramp-density")),0.2f);
+			rampDensity = (float) StrictMath.round(rampDensity *127)/127;
 		}
 
 		if (arguments.containsKey("reclaim-density")) {
-			RECLAIM_DENSITY = Float.parseFloat(arguments.get("reclaim-density"));
-			RECLAIM_DENSITY = (float) StrictMath.round(RECLAIM_DENSITY*127)/127;
+			reclaimDensity = Float.parseFloat(arguments.get("reclaim-density"));
+			reclaimDensity = (float) StrictMath.round(reclaimDensity *127)/127;
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java]: `127f` and spacing for binary operators and after commas.

File: src/java/generator/MapGenerator.java [hunk 5]
```diff
@@ -421,55 +447,55 @@ private static void parseMapName() {
...
-		RAMP_DENSITY = (float) RANDOM.nextInt(127)/127*0.2f;
-		RECLAIM_DENSITY = (float) RANDOM.nextInt(127)/127;
-		MEX_COUNT = 8 + 4/(SPAWN_COUNT/2) + RANDOM.nextInt(40/SPAWN_COUNT);
+		spawnCount = 6;
+		landDensity = StrictMath.min((random.nextInt(127)+13.0f)/127,1);
+		plateauDensity = (float) random.nextInt(127)/127*0.2f;
+		mountainDensity = (float) random.nextInt(127)/127*0.075f;
+		rampDensity = (float) random.nextInt(127)/127*0.2f;
+		reclaimDensity = (float) random.nextInt(127)/127;
+		mexCount = 8 + 4/(spawnCount /2) + random.nextInt(40/ spawnCount);
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java]: `127f`, also spacing

File: src/java/generator/MapGenerator.java [hunk 6]
```diff
@@ -404,13 +430,13 @@ private static void parseMapName() {
...
 				byte[] seedBytes = Base64.getDecoder().decode(seedString);
 				ByteBuffer seedWrapper = ByteBuffer.wrap(seedBytes);
-				SEED = seedWrapper.getLong();
+				seed = seedWrapper.getLong();
 			}
-			RANDOM = new Random(SEED);
+			random = new Random(seed);
 			randomizeOptions();
 		}
 		if (args.length>=6) {
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java:436]: space

File: src/java/generator/MapGenerator.java [hunk 7]
```diff
@@ -421,55 +447,55 @@ private static void parseMapName() {
...
-			RAMP_DENSITY = (float) optionBytes[4]/127*0.2f;
+			rampDensity = (float) optionBytes[4]/127*0.2f;
 		}
 		if (optionBytes.length>5){
-			RECLAIM_DENSITY = (float) optionBytes[5]/127;
+			reclaimDensity = (float) optionBytes[5]/127;
 		}
 		if (optionBytes.length>6){
-			MEX_COUNT = optionBytes[6];
+			mexCount = optionBytes[6];
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java]: spacing for all binary operators (divison, multiplication and comparators)

File: src/java/generator/MapGenerator.java [hunk 8]
```diff
@@ -59,17 +84,17 @@ public static void main(String[] args) throws ExecutionException, InterruptedExc
...
+		SCMap map = generator.generate();
+		System.out.println("Saving map to " + Paths.get(folderPath).toAbsolutePath() + "\\" + mapName.replace('/','^'));
+		System.out.println("Seed: "+ seed);
+		System.out.println("Land Density: "+ landDensity);
+		System.out.println("Plateau Density: "+ plateauDensity);
+		System.out.println("Mountain Density: "+ mountainDensity);
+		System.out.println("Ramp Density: "+ rampDensity);
+		System.out.println("Reclaim Density: "+ reclaimDensity);
+		System.out.println("Mex Count: "+ mexCount);
+		generator.save(folderPath, mapName.replace('/','^'), map);
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java]: spacing

File: src/java/generator/MapGenerator.java [hunk 9]
```diff
@@ -93,21 +118,80 @@ public void save(String folderName, String mapName, SCMap map) {
...
 		long startTime = System.currentTimeMillis();
-		final Random random = new Random(seed);
-		final int hydroCount = SPAWN_COUNT + random.nextInt(SPAWN_COUNT/2)*2;
-		final SCMap map = new SCMap(512, SPAWN_COUNT, MEX_COUNT*SPAWN_COUNT, hydroCount);
 
-		final ConcurrentBinaryMask land = new ConcurrentBinaryMask(16, random.nextLong(), "land");
-		final ConcurrentBinaryMask mountains = new ConcurrentBinaryMask(32, random.nextLong(), "mountains");
-		final ConcurrentBinaryMask plateaus = new ConcurrentBinaryMask(32, random.nextLong(), "plateaus");
-		final ConcurrentBinaryMask ramps = new ConcurrentBinaryMask(128, random.nextLong(), "ramps");
+		final int hydroCount = spawnCount + random.nextInt(spawnCount /2)*2;
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java]: spacing

File: src/java/generator/MapGenerator.java [hunk 10]
```diff
@@ -117,70 +201,87 @@ public SCMap generate(long seed) throws ExecutionException, InterruptedException
...
-		markerGenerator.generateSpawns(spawnsMask, 64);
-		markerGenerator.generateMexes(resourceMask);
-		markerGenerator.generateHydros(resourceMask);
+		cliffRockMask.randomize(.15f).intersect(rock).minus(plateaus).minus(mountains).intersect(land);
+		fieldStoneMask.randomize(reclaimDensity *.005f).enlarge(256).intersect(grass);
+		fieldStoneMask.enlarge(513).trimEdge(10);
+		treeMask.randomize(0.1f).inflate(1).cutCorners().acid(0.5f).enlarge(128).smooth(4).acid(0.5f);
+		treeMask.enlarge(256).intersect(grass);
+		treeMask.enlarge(513).deflate(5).trimEdge(3).fillCircle(256,256,96,false);
+		rockFieldMask.randomize(reclaimDensity *.005f).trimEdge(48).inflate(3).acid(.5f).intersect(land).minus(mountains);
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java]: spacing, is omitting the 0 in front of the decimal place a good idea?

File: src/java/generator/MapGenerator.java [hunk 11]
```diff
@@ -421,55 +447,55 @@ private static void parseMapName() {
...
-				(byte) (RAMP_DENSITY/0.2f*127),
-				(byte) (RECLAIM_DENSITY*127),
-				(byte) (MEX_COUNT)};
+		byte[] optionArray = {(byte) spawnCount,
+				(byte) (landDensity *127),
+				(byte) (plateauDensity /0.2f*127),
+				(byte) (mountainDensity /0.1f*127),
+				(byte) (rampDensity /0.2f*127),
+				(byte) (reclaimDensity *127),
+				(byte) (mexCount)};
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java]: spacing, `127f`

File: src/java/generator/MapGenerator.java (hunk 1 above)
Comment (Sheikah45) [reply src/java/generator/MapGenerator.java]: Makes sense. Done

File: src/java/generator/MapGenerator.java (hunk 2 above)
Comment (Sheikah45) [reply src/java/generator/MapGenerator.java]: Done

File: src/java/generator/MapGenerator.java (hunk 3 above)
Comment (Sheikah45) [reply src/java/generator/MapGenerator.java]: Done

File: src/java/generator/MapGenerator.java (hunk 4 above)
Comment (Sheikah45) [reply src/java/generator/MapGenerator.java]: Done

File: src/java/generator/MapGenerator.java (hunk 7 above)
Comment (Sheikah45) [reply src/java/generator/MapGenerator.java]: Done

File: src/java/generator/MapGenerator.java (hunk 9 above)
Comment (Sheikah45) [reply src/java/generator/MapGenerator.java]: Done

File: src/java/generator/MapGenerator.java (hunk 8 above)
Comment (Sheikah45) [reply src/java/generator/MapGenerator.java]: Done

File: src/java/generator/MapGenerator.java (hunk 10 above)
Comment (Sheikah45) [reply src/java/generator/MapGenerator.java]: Done and took out the zeros as it didnt seem to hurt.

File: src/java/generator/MapGenerator.java (hunk 11 above)
Comment (Sheikah45) [reply src/java/generator/MapGenerator.java]: Done

File: src/java/generator/MapGenerator.java (hunk 5 above)
Comment (Sheikah45) [reply src/java/generator/MapGenerator.java]: Done

File: src/java/generator/MapGenerator.java (hunk 6 above)
Comment (Sheikah45) [reply src/java/generator/MapGenerator.java:436]: Done

**Comment by Geosearchef** [comment]
I'd also squash the commits.

**Pull request merged**